src/linear_algebra/eigenspace.lean

```diff
@@ -0,0 +1,242 @@
+
+## Notations
+
+The expression `algebra_map α (β →ₗ[α] β)` appears very often, which is why we
+use `am` as a local notation for it.
+
+## References
+
+* [Sheldon Axler, *Down with determinants!*,
+  https://www.maa.org/sites/default/files/pdf/awards/Axler-Ford-1996.pdf][axler1996]
+* https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors
+
+## Tags
+
+eigenspace, eigenvector, eigenvalue, eigen
+-/
+
+universes u v w
+
+open vector_space principal_ideal_ring polynomial finite_dimensional
+
+variables {α : Type v} {β : Type w} [add_comm_group β]
+
+local notation `am` := algebra_map α (β →ₗ[α] β)
+
+/-- The subspace `eigenspace f μ` for a linear map `f` and a scalar `μ` consists of all vectors `x`
+    such that `f x = μ • x`. -/
+def eigenspace [field α] [vector_space α β] (f : β →ₗ[α] β) (μ : α) : submodule α β :=
+(f - am μ).ker
+
+/-- A nonzero element of an eigenspace is an eigenvector. -/
+def eigenvector [field α] [vector_space α β] (f : β →ₗ[α] β) (μ : α) (x : β) : Prop :=
+x ≠ 0 ∧ x ∈ eigenspace f μ
+
+/-- A scalar `μ` is an eigenvalue for a linear map `f` if there are nonzero vectors `x`
+    such that `f x = μ • x`. -/
+def eigenvalue [field α] [vector_space α β] (f : β →ₗ[α] β) (a : α) : Prop :=
+eigenspace f a ≠ ⊥.
+
+lemma eigenspace_div [field α] [vector_space α β] (f : β →ₗ[α] β) (a b : α) (hb : b ≠ 0):
+  eigenspace f (a / b) = (b • f - am a).ker :=
+calc
+  eigenspace f (a / b) = eigenspace f (b⁻¹ * a) : by { dsimp [(/)], rw mul_comm }
+  ... = (f - (b⁻¹ * a) • linear_map.id).ker : by refl
+  ... = (f - b⁻¹ • a • linear_map.id).ker : by rw smul_smul
+  ... = (f - b⁻¹ • am a).ker : by refl
+  ... = (b • (f - b⁻¹ • am a)).ker : by rw linear_map.ker_smul _ b hb
+  ... = (b • f - am a).ker : by rw [smul_sub, smul_inv_smul' hb]
+
+lemma eigenspace_eval₂_polynomial_degree_1 [field α] [vector_space α β]
+  (f : β →ₗ[α] β) (q : polynomial α) (hq : degree q = 1) :
+  eigenspace f (- q.coeff 0 / q.leading_coeff) = (eval₂ am f q).ker :=
+calc
+  eigenspace f (- q.coeff 0 / q.leading_coeff) = (q.leading_coeff • f - am (- q.coeff 0)).ker
+    : by { rw eigenspace_div, intro h, rw leading_coeff_eq_zero_iff_deg_eq_bot.1 h at hq, cases hq }
+  ... = (eval₂ am f (C q.leading_coeff * X + C (q.coeff 0))).ker
+    : by { rw C_mul', simpa [algebra_map, algebra.to_ring_hom] }
+  ... = (eval₂ am f q).ker
+     : by { congr, apply (eq_X_add_C_of_degree_eq_one hq).symm }
+
+lemma mem_eigenspace_iff [field α] [vector_space α β]
+ {f : β →ₗ[α] β} {μ : α} {x : β} : x ∈ eigenspace f μ ↔ f x = μ • x :=
+by rw [eigenspace, linear_map.mem_ker, linear_map.sub_apply, module.endomorphism_algebra_map_apply2,
+   sub_eq_zero]
+
+lemma ne_0_of_mem_factors {α : Type v} [field α] {p q : polynomial α}
+  (hp : p ≠ 0) (hq : q ∈ factors p) : q ≠ 0 :=
+begin
+  intro h_q_eq_0,
+  rw h_q_eq_0 at hq,
+  apply hp ((associated_zero_iff_eq_zero p).1 _),
+  rw ←multiset.prod_eq_zero hq,
+  apply (factors_spec p hp).2
+end
```

Comment:
This should hold for any `p q : R` where `R` is a principal ideal domain, so could you generalize it and move it to `principal_ideal_domain.lean`?

I think the proof can be shortened to essentially `irreducible.ne_zero (factors_spec p hp).1`.